optimum/utils/preprocessing/base.py

```diff
@@ -1,101 +1,224 @@
+
+    def create_dataset_processing_func(
+        self, data_keys: Dict[str, str], ref_keys: Optional[List[str]] = None
+    ) -> Callable[[Dict[str, Any]], Dict[str, Any]]:
+        return functools.partial(self.dataset_processing_func, data_keys=data_keys, ref_keys=ref_keys)
+
+    def prepare_dataset(
+        self,
+        dataset: Union[DatasetDict, Dataset],
+        data_keys: Dict[str, str],
+        ref_keys: Optional[List[str]] = None,
+        split: Optional[str] = None,
+    ) -> Union[DatasetDict, Dataset]:
+        if isinstance(dataset, Dataset) and split is not None:
+            raise ValueError("A Dataset and a split name were provided, but splits are for DatasetDict.")
+        elif split is not None:
+            dataset = dataset[split]
+        return dataset.map(self.create_dataset_processing_func(data_keys, ref_keys))
+
+    @abstractmethod
+    def try_to_guess_data_keys(self, column_names: List[str]) -> Optional[Dict[str, str]]:
+        raise NotImplementedError("This method must be implemented in subclasses.")
+
+    @abstractmethod
+    def try_to_guess_ref_keys(self, column_names: List[str]) -> Optional[List[str]]:
+        raise NotImplementedError("This method must be implemented in subclasses.")
+
+    def load_dataset(
+        self,
+        path: str,
+        data_keys: Optional[Dict[str, str]] = None,
+        ref_keys: Optional[List[str]] = None,
+        only_keep_necessary_columns: bool = False,
+        load_smallest_split: bool = False,
+        num_samples: Optional[int] = None,
+        shuffle: bool = False,
+        **load_dataset_kwargs,
+    ) -> Union[DatasetDict, Dataset]:
+        dataset = datasets_load_dataset(path, **load_dataset_kwargs)
+
+        if isinstance(dataset, DatasetDict) and load_smallest_split:
+            split = load_dataset_kwargs.get("split", None)
+            if split is not None:
+                raise ValueError(
+                    f"A split name was provided ({split}) but load_smallest_split is True, use either one or the other."
+                )
+            smallest_split = min(dataset.items(), key=lambda item: item[1].num_rows)[0]
+            dataset = dataset[smallest_split]
+
+        if shuffle:
+            dataset = dataset.shuffle()
+
+        if isinstance(dataset, Dataset) and num_samples is not None:
+            if num_samples > dataset.num_rows:
+                raise ValueError(
+                    f"There are only {dataset.num_rows} examples in the dataset, but it was requested to select "
+                    f"{num_samples} examples."
+                )
+            dataset = dataset.select(range(num_samples))
+
+        column_names = dataset.column_names
+        if isinstance(column_names, dict):
+            column_names = list(set(itertools.chain.from_iterable(column_names.values())))
+
+        if data_keys is None:
+            data_keys = self.try_to_guess_data_keys(column_names)
```

Comment:
A warning should be raised here I think - just to avoid unexpected behavior

Reply:
Done